src/hash/blake/mod.rs

```diff
@@ -0,0 +1,226 @@
+use crate::{
+    ByteReader, ByteWriter, Deserializable, DeserializationError, Digest, ElementHasher, Felt,
+    FieldElement, HashFn, Serializable, StarkField,
+};
+use core::{
+    mem::{size_of, transmute, transmute_copy},
+    ops::Deref,
+    slice::from_raw_parts,
+};
+
+#[cfg(test)]
+mod tests;
+
+// BLAKE3 N-BIT OUTPUT
+// ================================================================================================
+
+/// N-bytes output of a blake3 function.
+///
+/// Note: `N` can't be greater than `32`
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct Blake3Digest<const N: usize>([u8; N]);
+
+impl<const N: usize> Default for Blake3Digest<N> {
+    fn default() -> Self {
+        Self([0; N])
+    }
+}
+
+impl<const N: usize> Deref for Blake3Digest<N> {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<const N: usize> Serializable for Blake3Digest<N> {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u8_slice(&self.0);
+    }
+}
+
+impl<const N: usize> Deserializable for Blake3Digest<N> {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        source.read_u8_array().map(Self)
+    }
+}
+
+impl<const N: usize> Digest for Blake3Digest<N> {
+    fn as_bytes(&self) -> [u8; 32] {
+        // compile-time assertion
+        assert!(N <= 32);
```

Comment:
I think it will more informative to the user if we also throw a message here when the assertion fails.

Reply:
Fixed!